### PR TITLE
SNOW-787955: resolve namespace collision for udf/stored proc imports 

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

  SNOW-787955

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.
    The original PR for SnowML here. https://github.com/snowflakedb/snowml/pull/589/files

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: tests/integ/test_udf.py [hunk 1]
```diff
@@ -720,6 +720,74 @@ def plus4_then_mod5(x):
...
+
+            from snowflake.snowpark.functions import udf  # noqa: F401
+
+            return f(x)
+
+        session.add_import(str(py_file), import_path="snowflake.task.test_udf_file")
+
+        df = session.range(-5, 5).to_df("a")
+
+        plus4_then_mod5_udf = udf(
```
Comment (sfc-gh-wzhao) [inline tests/integ/test_udf.py:779]: nit: This `plus4_then_mod5_udf` name could be changed as it is not a udf that addes 4 and then mods 5.

File: tests/unit/scala/test_utils_suite.py [hunk 2]
```diff
@@ -147,36 +147,35 @@ def check_zip_files_and_close_stream(input_stream, expected_files):
...
     with zip_file_or_directory_to_stream(
         test_files.test_udf_py_file,
         leading_path=os.path.dirname(test_files.test_udf_directory),
     ) as stream:
-        check_zip_files_and_close_stream(stream, ["test_udf_dir/test_udf_file.py"])
+        check_zip_files_and_close_stream(
+            stream, ["test_udf_dir/", "test_udf_dir/test_udf_file.py"]
+        )
 
     with zip_file_or_directory_to_stream(
```
Comment (sfc-gh-wzhao) [inline tests/unit/scala/test_utils_suite.py:161]: nit: it seems that this test case is now identical to the one above.

**Review by sfc-gh-wzhao — APPROVED** [review]

File: tests/unit/scala/test_utils_suite.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply tests/unit/scala/test_utils_suite.py:161]: good catch, done

File: tests/integ/test_udf.py (hunk 1 above)
Comment (sfc-gh-jdu) [reply tests/integ/test_udf.py]: done

File: src/snowflake/snowpark/_internal/utils.py [hunk 3]
```diff
@@ -319,6 +319,11 @@ def zip_file_or_directory_to_stream(
     with zipfile.ZipFile(
         input_stream, mode="w", compression=zipfile.ZIP_DEFLATED
     ) as zf:
+        cur_path = os.path.dirname(path)
+        while os.path.realpath(cur_path) != os.path.realpath(start_path):
+            zf.writestr(f"{os.path.relpath(cur_path, start_path)}/", "")
```
[the remark this answers is not in the document]
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/utils.py:324]: @sfc-gh-jdu : Let's add a comment about this?

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/_internal/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/utils.py:326]: added

**Pull request merged**